File: backend/src/main/java/com/associazionedisabili/controller/EventController.java
package com.associazionedisabili.controller;

import com.associazionedisabili.dto.event.EventRequest;
import com.associazionedisabili.dto.event.EventResponse;
import com.associazionedisabili.model.Event;
import com.associazionedisabili.service.CalendarInviteService;
import com.associazionedisabili.service.EventService;
import jakarta.validation.Valid;
import java.time.LocalDateTime;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/events")
@RequiredArgsConstructor
public class EventController {

 private final EventService service;
 private final CalendarInviteService calendarInviteService;

 @PostMapping
 public ResponseEntity<EventResponse> create(@Valid @RequestBody EventRequest request) {
  return ResponseEntity.ok(service.create(request));
 }

 @GetMapping
 public List<EventResponse> list() {
  return service.list();
 }

 @GetMapping("/filter")
 public List<EventResponse> getEventsByDateRange(
  @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime start,
  @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime end
 ) {
  return service.getEventsByDateRange(start, end);
 }

 @GetMapping("/upcoming")
 public List<EventResponse> getUpcomingEvents(
  @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime from
 ) {
  LocalDateTime fromDate = from != null ? from : LocalDateTime.now();
  return service.getUpcomingEvents(fromDate);
 }

 @GetMapping("/{id}")
 public EventResponse getById(@PathVariable Long id) {
  return service.getById(id);
 }

 @PutMapping("/{id}")
 public EventResponse update(@PathVariable Long id, @Valid @RequestBody EventRequest request) {
  return service.update(id, request);
 }

 @DeleteMapping("/{id}")
 public ResponseEntity<Void> delete(@PathVariable Long id) {
  service.delete(id);
  return ResponseEntity.noContent().build();
 }

 @GetMapping("/{id}/calendar")
 public ResponseEntity<ByteArrayResource> downloadCalendar(@PathVariable Long id) {
  Event event = service.getEntityById(id);
  ByteArrayResource resource = new ByteArrayResource(calendarInviteService.buildEventInvite(event));
  return ResponseEntity.ok()
   .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + calendarInviteService.buildFileName(event) + "\"")
   .contentType(MediaType.parseMediaType("text/calendar"))
   .body(resource);
 }
}

File: backend/Controllers/EventController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using SoccerDreamFermana.Backend.Dtos;
using SoccerDreamFermana.Backend.Services;

namespace SoccerDreamFermana.Backend.Controllers;

[ApiController]
[Route("api/events")]
public sealed class EventController(EventService eventService, CalendarInviteService calendarInviteService) : ControllerBase
{
    [HttpPost]
    [Authorize(Roles = "ADMIN")]
    public Task<EventResponse> Create(EventRequest request, CancellationToken cancellationToken)
    {
        return eventService.CreateAsync(request, cancellationToken);
    }

    [HttpGet]
    [AllowAnonymous]
    public Task<IReadOnlyList<EventResponse>> List([FromQuery] string? tipo, CancellationToken cancellationToken)
    {
        return eventService.ListAsync(tipo, cancellationToken);
    }

    [HttpGet("filter")]
    [AllowAnonymous]
    public Task<IReadOnlyList<EventResponse>> GetEventsByDateRange(
        [FromQuery] DateTime start,
        [FromQuery] DateTime end,
        [FromQuery] string? tipo,
        CancellationToken cancellationToken
    )
    {
        return eventService.GetEventsByDateRangeAsync(start, end, tipo, cancellationToken);
    }

    [HttpGet("upcoming")]
    [AllowAnonymous]
    public Task<IReadOnlyList<EventResponse>> GetUpcomingEvents(
        [FromQuery] DateTime? from,
        [FromQuery] string? tipo,
        CancellationToken cancellationToken
    )
    {
        return eventService.GetUpcomingEventsAsync(from ?? DateTime.Now, tipo, cancellationToken);
    }

    [HttpGet("{id:long}")]
    [AllowAnonymous]
    public Task<EventResponse> GetById(long id, CancellationToken cancellationToken)
    {
        return eventService.GetByIdAsync(id, cancellationToken);
    }

    [HttpPut("{id:long}")]
    [Authorize(Roles = "ADMIN")]
    public Task<EventResponse> Update(long id, EventRequest request, CancellationToken cancellationToken)
    {
        return eventService.UpdateAsync(id, request, cancellationToken);
    }

    [HttpDelete("{id:long}")]
    [Authorize(Roles = "ADMIN")]
    public async Task<IActionResult> Delete(long id, CancellationToken cancellationToken)
    {
        await eventService.DeleteAsync(id, cancellationToken);
        return NoContent();
    }

    [HttpGet("{id:long}/calendar")]
    [AllowAnonymous]
    public async Task<IActionResult> DownloadCalendar(long id, CancellationToken cancellationToken)
    {
        var item = await eventService.GetEntityByIdAsync(id, cancellationToken);
        return File(
            calendarInviteService.BuildEventInvite(item),
            "text/calendar",
            calendarInviteService.BuildFileName(item)
        );
    }
}
